module/zfs/zap.c

```diff
@@ -969,11 +975,323 @@ fzap_remove(zap_name_t *zn, dmu_tx_t *tx)
+ * 2. Check if the sibling leaf is empty.
+ * 3. If sibling bit of the initial leaf is not 0, release it.
+ *    In order to avoid deadlock, we always dereference leaves in the same order
+ *    - the leaf with sibling bit 0 first, then the leaf with sl_bit 1.
+ * 4. Upgrade zapdir lock to WRITER (once).
+ * 5. Deref leaves if needed.
+ * 6. Recheck both leaves if required.
+ * 7. Update ptrtbl pointers of the leaf with sl_bit 1 (sibling leaf)to point to
+ *    the leaf with sl_bit 0.
+ * 8. Free disk space of the sibling leaf (dmu_free_range).
+ * 9. Update prefix and prefix_len of the remaining leaf.
+ *    Now we have some new leaf, so we repeat the removing procedure for it.
+ */
+static int
+zap_shrink(zap_name_t *zn, zap_leaf_t *l, dmu_tx_t *tx)
+{
+	zap_t *zap = zn->zn_zap;
+	int64_t zt_shift = zap_f_phys(zap)->zap_ptrtbl.zt_shift;
+	uint64_t hash = zn->zn_hash;
+	uint64_t prefix = zap_leaf_phys(l)->l_hdr.lh_prefix;
+	uint64_t prefix_len = zap_leaf_phys(l)->l_hdr.lh_prefix_len;
+	int nshrunk = 0;
+	int err = 0;
+
+	ASSERT3U(zap_leaf_phys(l)->l_hdr.lh_nentries, ==, 0);
+	ASSERT3U(prefix_len, <=, zap_f_phys(zap)->zap_ptrtbl.zt_shift);
+	ASSERT(RW_LOCK_HELD(&zap->zap_rwlock));
+	ASSERT3U(ZAP_HASH_IDX(hash, prefix_len), ==, prefix);
+
+	boolean_t writer = B_FALSE;
+
+	// XXX - DTRACE_PROBE(zap__shrink__start);
+
+	/*
+	 * To avoid deadlock always deref leaves in same order - l0, l1.
+	 * XXX: ???
+	 */
+	while (prefix_len) {
+		zap_leaf_t *sl;
+		int64_t prefix_diff = zt_shift - prefix_len;
+		uint64_t sl_prefix = prefix ^ 1;
+		uint64_t sl_hash = ZAP_PREFIX_HASH(sl_prefix, prefix_len);
+		int slbit = prefix & 1;
+
+		ASSERT3U(zt_shift, ==, zap_f_phys(zap)->zap_ptrtbl.zt_shift);
+		ASSERT3S(zap_leaf_phys(l)->l_hdr.lh_nentries, ==, 0);
+
+
+		/*
+		 * Check if there is sibling by reading corresponding tbl ptrs.
+		 */
+		if (zap_check_leaf_by_ptrtbl(zap, sl_prefix, prefix_len) == 0)
+			break;
+
+		/*
+		 * This is l1 leaf - have to unlock before locking the sibling.
+		 */
+		if (slbit == 1) {
+			zap_put_leaf(l);
+			l = NULL;
+		}
+
+		/*
+		 * Deref sibling leaf and check if it is empty.
+		 */
+		if ((err = zap_deref_leaf(zap, sl_hash, tx, RW_READER,
+		    &sl)) != 0)
+			break;
+
+		ASSERT3U(ZAP_HASH_IDX(sl_hash, prefix_len), ==, sl_prefix);
+
+		if (zap_leaf_phys(sl)->l_hdr.lh_prefix_len != prefix_len) {
+			ASSERT(0); // shouldn't happen, see check sibl. above
+		}
+
+		/*
+		 * Check if we have an empty sibling.
+		 */
+		if (zap_leaf_phys(sl)->l_hdr.lh_prefix_len != prefix_len ||
+		    zap_leaf_phys(sl)->l_hdr.lh_nentries != 0) {
+			zap_put_leaf(sl);
+			break;
+		}
+
+		zap_put_leaf(sl);
+
+		/*
+		 * We need writer lock to modify ZAP ptrtbl.
+		 */
+		if (!writer && (writer = zap_tryupgradedir(zap, tx)) == 0) {
+			/* We failed to upgrade */
+			if (l != NULL) {
+				zap_put_leaf(l);
+				l = NULL;
+			}
+
+			/*
+			 * Instead of calling zap_unlockdir(); zap_lockdir();
+			 * we do it in more optimized way.
```

Comment:
This seems like an unnecessary optimization, given that we don't make this optimization in `zap_expand_leaf()`, which is probably called much more often than this.